pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -236,11 +248,41 @@ type externalServiceValue struct {
 	Service *alpha1.CDAPExternalServiceSpec
 }
 
+// Struct containing data for templatization using UpgradeJobSpec
+type upgradeValue struct {
+	templateBaseValue
+	Job *UpgradeJobSpec
+}
+
+// UpgradeJobSpec defines the specification for the upgrade job
+type UpgradeJobSpec struct {
+	Image        string `json:"image,omitempty"`
+	JobName      string `json:"jobName,omitempty"`
+	HostName     string `json:"hostName,omitempty"`
+	BackoffLimit int32  `json:"backoffLimit, omitempty"`
+}
+
 // Returns the config map name for the given configuration type
 func getConfigName(r *alpha1.CDAPMaster, confType string) string {
 	return fmt.Sprintf("cdap-%s-%s", r.Name, confType)
 }
 
+// Gets the set of resources for the pre-upgrade job.
+func getUpgradeResources(s *UpgradeJobSpec, rsrclabels map[string]string) ([]reconciler.Object, error) {
+	ngdata := &upgradeValue{Job: s}
+	ngdata.Labels = rsrclabels
+	item, err := k8s.ObjectFromFile(templateDir+upgradeJobTemplate, ngdata, &batchv1.JobList{})
+	if err != nil {
+		return nil, err
+	}
+	return append([]reconciler.Object{}, *item), nil
+}
+
+// Gets the name of the upgrade job based on start time.
+func getUpgradeJobName(startTime int64) string {
+	return "upgrade-job" + strconv.FormatInt(startTime, 10)
```

Comment:
The job name should be prefixed with `"cdap-" + masterSpec.Name` to make it clear which CDAP instance the job is for. See the setTemplateValue method https://github.com/cdapio/cdap-operator/pull/9/files#diff-0c39caca86fbd994c47c05c67732106aR185

Also, the master spec name is unique without a K8s namespace, so maybe we don't need the startTimeNano to make the job name unique.

Reply:
What if an upgrade is retried by the user and the job has not been removed by ttlSeconds yet? Won't the name field still be the same?